Fixed redis start operation if replication sync takes > 20 seconds

elconas opened the pull request:
Current Resource Agent wait 20 seconds to initial sync to finish. 

If the sync takes longer then 20 seconds (can be if huge amount of data is synced), the start fails, altought fully OK. This fix waits for the sync to finish forever. meta op start timeut=xxx should set an appropriate value for the start to timeout.

dmuhamedagic commented:
Ad-hoc timeouts, such as this one:

 for (( c=1; c <= 20; c++ )) 

should be removed. The timeout is _always_ handled by the resource manager.

I'm not sure whether this extra test for replication is necessary at all.

davidvossel commented:
There are two of these for loops. one in "start" and one in "promote" (i really hope i didn't add these because they are terrible).  If we convert these two loops to infinite loops, then the resource manager will be able to control the timeout properly.

elconas commented:
Background (from my tests): 

Redis master_link_status will only come "up" when the SYNC with the master has completed. This can take an arbitraty time (depending on the amound of data) and should only be parametrized by the start operation timeout by the administrator, not by this resource agent code. The start operation should not complete until full sync is restored (I guess?) as we could also skip the check for "up" in case the sync is running, however then we could failback to partially synced node  which is not good. So completing the "start" when "sync is finished" sounds more stable in terms of data integrity.

davidvossel commented:
@elconas yep, i agree.  start = up+syned

elconas commented:
Hello, 

1) "start" loop

This loop seems to be infinite (while true; do) , which is ok - or am I missing something ?

2) "demote" loop 

I removed the "c<20" and removed the slave status.

Regards, 
Robert

davidvossel commented:
looks good. thanks!
